src/core/service.c

```diff
@@ -2684,33 +2690,174 @@ static void service_enter_reload_by_notify(Service *s) {
 
         service_set_state(s, SERVICE_RELOAD_NOTIFY);
 
-        /* service_enter_reload_by_notify is never called during a reload, thus no loops are possible. */
+        /* service_enter_reload_notify is never called during a reload, thus no loops are possible. */
         r = manager_propagate_reload(UNIT(s)->manager, UNIT(s), JOB_FAIL, &error);
         if (r < 0)
                 log_unit_warning(UNIT(s), "Failed to schedule propagation of reload, ignoring: %s", bus_error_message(&error, r));
 }
 
-static void service_enter_reload(Service *s) {
-        bool killed = false;
+static void service_enter_reload_signal(Service *s) {
         int r;
 
         assert(s);
 
         service_unwatch_control_pid(s);
         s->reload_result = SERVICE_SUCCESS;
 
-        usec_t ts = now(CLOCK_MONOTONIC);
-
-        if (s->type == SERVICE_NOTIFY_RELOAD && pidref_is_set(&s->main_pid)) {
+        if (pidref_is_set(&s->main_pid)) {
                 r = pidref_kill_and_sigcont(&s->main_pid, s->reload_signal);
                 if (r < 0) {
                         log_unit_warning_errno(UNIT(s), r, "Failed to send reload signal: %m");
                         goto fail;
                 }
 
-                killed = true;
+                r = service_arm_timer(s, /* relative= */ true, s->timeout_start_usec);
+                if (r < 0) {
+                        log_unit_warning_errno(UNIT(s), r, "Failed to install timer: %m");
+                        goto fail;
+                }
+
+                service_set_state(s, SERVICE_RELOAD_SIGNAL);
+        } else
+                service_enter_running(s, SERVICE_SUCCESS);
+
+        return;
+
+fail:
+        s->reload_result = SERVICE_FAILURE_RESOURCES;
+        service_enter_running(s, SERVICE_SUCCESS);
+}
+
+static bool service_should_reload_extensions(Service *s) {
+        int r;
+
+        assert(s);
+
+        if (!pidref_is_set(&s->main_pid)) {
+                log_unit_debug(UNIT(s), "Not reloading extensions for service without main PID.");
+                return false;
         }
 
+        r = exec_context_has_vpicked_extensions(&s->exec_context);
+        if (r < 0)
+                log_unit_warning_errno(UNIT(s), r, "Failed to determine if service should reload extensions, assuming false: %m");
+        if (r == 0)
+                log_unit_debug(UNIT(s), "Service has no extensions to reload.");
+        if (r <= 0)
+                return false;
+
+        // TODO: Add support for user services, which can use ExtensionDirectories= + notify-reload.
+        // For now, skip for user services.
+        if (UNIT(s)->manager->runtime_scope != RUNTIME_SCOPE_SYSTEM) {
```

Comment:
!MANAGER_IS_SYSTEM()

Reply:
not addressed?

Reply:
I have not pushed yet, need to find time to get back for one of the requested changes, but I've marked as resolved what I fixed in my local branch so that I don't forget